src/psyclone/tests/psyir/transformations/inline_trans_test.py

```diff
@@ -772,6 +772,172 @@ def test_apply_multi_function(fortran_reader, fortran_writer, tmpdir):
+    routine = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    inline_trans.apply(routine)
+    output = fortran_writer(psyir)
+    expected = (
+        f"{indent}subroutine run_it()\n"
+        f"{indent}  real :: a\n\n"
+        f"{indent}  a = 2.0 * a\n\n"
+        f"{indent}end subroutine run_it\n")
+    assert expected in output
+    if "use dummy" not in output:
+        # Compilation will not work with "use dummy" as there is no
+        # mod file.
+        assert Compile(tmpdir).string_compiles(output)
+
+
+@pytest.mark.parametrize("use1, use2", [
+    ("use inline_mod, only : sub\n", ""), ("use inline_mod\n", ""),
+    ("", "use inline_mod, only : sub\n"), ("", "use inline_mod\n")])
+def test_apply_container_subroutine(
+        fortran_reader, fortran_writer, tmpdir, use1, use2):
+    '''Test the apply method works correctly when the routine to be
+    inlined is in a different container and is within a module (so
+    a use statement is required).
+
+    '''
+    code = (
+        f"module inline_mod\n"
+        f"contains\n"
+        f"  subroutine sub(x)\n"
+        f"    real, intent(inout) :: x\n"
+        f"    x = 2.0*x\n"
+        f"  end subroutine sub\n"
+        f"end module inline_mod\n"
+        f"module test_mod\n"
+        f"{use1}"
+        f"contains\n"
+        f"  subroutine run_it()\n"
+        f"    {use2}"
+        f"    real :: a\n"
+        f"    call sub(a)\n"
+        f"  end subroutine run_it\n"
+        f"end module test_mod\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    inline_trans.apply(routine)
+    output = fortran_writer(psyir)
+    assert (
+        "    real :: a\n\n"
+        "    a = 2.0 * a\n\n"
+        "  end subroutine run_it" in output)
+    assert Compile(tmpdir).string_compiles(output)
+
+
+def test_apply_container_subroutine_error(fortran_reader):
+    '''Test the apply method does not find the Routine if the Call is
+    imported but the Routine is not in a Container.
+
+    '''
+    code = (
+        "module test_mod\n"
+        "use inline_mod, only : sub\n"
+        "contains\n"
+        "  subroutine run_it()\n"
+        "    real :: a\n"
+        "    call sub(a)\n"
+        "  end subroutine run_it\n"
+        "end module test_mod\n"
+        "subroutine sub(x)\n"
+        "  real, intent(inout) :: x\n"
+        "  x = 2.0*x\n"
+        "end subroutine sub\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    with pytest.raises(TransformationError) as info:
+        inline_trans.apply(routine)
+    assert ("Failed to find the source for routine 'sub' and therefore "
+            "cannot inline it." in str(info.value))
+
+
+def test_apply_call_container_error(fortran_reader):
+    '''Test the apply method does not find the Routine if the Call is
+    not imported but the Routine is in a Container.
```

Comment:
To me this implies that the Routine in the Container is the one we want whereas it isn't. Might be better to say something like "in the presence of a routine of the same name in another module".

Reply:
Also, probably better to test the utility routine directly here.

Reply:
Rewritten tests. Hopefully both points have been addressed.